platforms/jvm/plugins-application/src/main/java/org/gradle/jvm/application/tasks/CreateStartScripts.java

```diff
@@ -147,85 +157,66 @@ protected JavaModuleDetector getJavaModuleDetector() {
-            return null;
-        }
-
-        return GUtil.toConstant(getApplicationName()) + "_OPTS";
+    @ReplacesEagerProperty
+    public Property<String> getOptsEnvironmentVar() {
+        return optsEnvironmentVar;
     }
 
     /**
      * The environment variable to use to control exit value (Windows only).
      */
-    @Nullable
     @Optional
     @Input
-    public String getExitEnvironmentVar() {
-        if (GUtil.isTrue(exitEnvironmentVar)) {
-            return exitEnvironmentVar;
-        }
-
-        if (!GUtil.isTrue(getApplicationName())) {
-            return null;
-        }
-
-        return GUtil.toConstant(getApplicationName()) + "_EXIT_CONSOLE";
+    @ReplacesEagerProperty
+    public Property<String> getExitEnvironmentVar() {
+        return exitEnvironmentVar;
     }
 
     /**
      * Returns the full path to the Unix script. The target directory is represented by the output directory, the file name is the application name without a file extension.
+     * TODO: This should be Provider[RegularFile], but we don't support such upgrade with @ReplacesEagerProperty
      */
     @Internal
-    public File getUnixScript() {
-        return new File(getOutputDir(), getApplicationName());
+    @ReplacesEagerProperty(replacedAccessors = @ReplacedAccessor(value = GETTER, name = "getUnixScript"))
+    public RegularFileProperty getUnixScript() {
+        return getObjectFactory().fileProperty().value(
+            getOutputDir().zip(getApplicationName(), Directory::file)
+        );
     }
 
     /**
      * Returns the full path to the Windows script. The target directory is represented by the output directory, the file name is the application name plus the file extension .bat.
+     * TODO: This should be Provider[RegularFile], but we don't support such upgrade with @ReplacesEagerProperty
      */
     @Internal
-    public File getWindowsScript() {
-        return new File(getOutputDir(), getApplicationName() + ".bat");
+    @ReplacesEagerProperty(replacedAccessors = @ReplacedAccessor(value = GETTER, name = "getWindowsScript"))
+    public RegularFileProperty getWindowsScript() {
+        return getObjectFactory().fileProperty().value(
+            getOutputDir().zip(getApplicationName(), (outputDir, applicationName) -> outputDir.file(applicationName + ".bat"))
+        );
     }
 
     /**
      * The directory to write the scripts into.
      */
     @OutputDirectory
-    @Nullable
-    public File getOutputDir() {
+    @ReplacesEagerProperty
+    public DirectoryProperty getOutputDir() {
         return outputDir;
     }
 
-    public void setOutputDir(@Nullable File outputDir) {
-        this.outputDir = outputDir;
-    }
-
     /**
      * The directory to write the scripts into in the distribution.
      * @since 4.5
      */
     @Input
-    public String getExecutableDir() {
+    @ReplacesEagerProperty
+    public Property<String> getExecutableDir() {
         return executableDir;
     }
```

Comment:
❓ How mach harder would it be to upgrade this to a `DirectoryProperty`?

Reply:
From bytecode backward compatibility that is not a problem. But that makes is also "source incompatible" change. So I would avoid doing these changes. I think we should discuss this more